Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs

```diff
@@ -0,0 +1,614 @@
+        /// </returns>
+        /// <param name="indexStart">The index of the row where the method should begin to look for the next <see cref="T:System.Windows.Forms.DataGridViewRow"/>.</param><param name="includeFilter">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="indexStart"/> is less than -1.</exception><exception cref="T:System.ArgumentException"><paramref name="includeFilter"/> is not a valid bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values.</exception>
+        public int GetNextRow(int indexStart, DataGridViewElementStates includeFilter)
+        {
+            Contract.Requires(indexStart < -1);
+            Contract.Requires(
+                !((includeFilter &
+                  ~(DataGridViewElementStates.Displayed | DataGridViewElementStates.Frozen |
+                    DataGridViewElementStates.ReadOnly | DataGridViewElementStates.Resizable |
+                    DataGridViewElementStates.Selected | DataGridViewElementStates.Visible)) !=
+                DataGridViewElementStates.None));
+            Contract.Ensures(Contract.Result<int>() >= -1);
+            return default(int);
+        }
+
+        /// <summary>
+        /// Returns the index of the next <see cref="T:System.Windows.Forms.DataGridViewRow"/> that meets the specified inclusion and exclusion criteria.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// The index of the next <see cref="T:System.Windows.Forms.DataGridViewRow"/> that has the attributes specified by <paramref name="includeFilter"/>, and does not have the attributes specified by <paramref name="excludeFilter"/>; -1 if no row is found.
+        /// </returns>
+        /// <param name="indexStart">The index of the row where the method should begin to look for the next <see cref="T:System.Windows.Forms.DataGridViewRow"/>.</param><param name="includeFilter">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values.</param><param name="excludeFilter">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="indexStart"/> is less than -1.</exception><exception cref="T:System.ArgumentException">One or both of the specified filter values is not a valid bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values.</exception>
+        public int GetNextRow(int indexStart, DataGridViewElementStates includeFilter, DataGridViewElementStates excludeFilter)
+        {
+            Contract.Requires(indexStart < -1);
+            Contract.Requires(
+                !((includeFilter &
```

Comment:
The same for everyone. But I think it would be useful to extract this stuff into method... Too verbose and error prone...